src/utils/buildMetadata.ts

```diff
@@ -166,13 +153,198 @@ export function buildMetadata(sheets: Sheet[], defaultCC: string) {
         optionSets,
         trackedEntityAttributes,
         trackedEntityTypes,
-        //programs,
+        programSections: buildprogramSections(sheets),
+        programs: buildPrograms(sheets),
+        programStages: buildProgramStages(sheets),
+        programStageSections: buildProgramStageSections(sheets),
         programRules: buildProgramRules(sheets),
         programRuleActions: buildProgramRuleActions(sheets),
         programRuleVariables: buildProgramRuleVariables(sheets),
     };
 }
 
+function buildPrograms(sheets: Sheet[]) {
+    const get = (name: string) => getItems(sheets, name);
+
+    const programs = get("programs");
+    const pSections = get("programSections");
+    const pStages = get("programStages");
+    const trackedEntityTypes = get("trackedEntityTypes");
+    const categoryCombos = get("categoryCombos");
+
+    return programs.map(program => {
+        const trackedEntityType = {
+            id: getByName(trackedEntityTypes, program.trackedEntityType)?.id
+        };
+
+        const programStages = pStages.filter((programStages) => {
+            return programStages?.program === program.name;
+        }).map(programStages => ({ id: programStages.id }));
+
+        replaceById(program, "categoryCombo", categoryCombos);
```

Comment:
The tricky part here is that `replaceById` modifies its argument (it changes in `program` the field `categoryCombo`), but we don't want `buildPrograms()` (which promises to "build" but doesn't say anything about changing the things we pass to it) to do that.

The way `replaceById()` can be used safely is by first making a copy of the object, and then modify things in the copy, which we will finally return as the metadata. You can see an example in `buildProgramRuleActions()`:

```ts
let data = { ...action } as MetadataItem; // copy that we will modify
replaceById(data, "programRule", rules);
[...]
return data;
```

If you do something like that (that is, starting with `let data = { ...program } as MetadataItem;`), then you can safely use it and also replace the previous `map`, so lines 182 to 184 could look like:

```ts
replaceById(data, "programStage", programStages);
replaceById(data, "categoryCombo", categoryCombos);
```

Again, this idea may be used in other points!

Reply:
Related to this, when later on you use `makeSortOrder()` (which is actually *adding* rather than *making*), the same warning applies about overwriting the variables that you pass to the function!

Reply:
As said above, programStages can have multiple Ids, agree about copying the object.